wbobeirne/joule-extension
# Ask for password only once

Closes #96 

### Description

Adds a little background script to cache the password after you've entered it. Before asking for password, attempt to fetch from that cache. This method should be completely safe for a few reasons:

1. The password is stored in memory, not using any long-term storage. Closing the browser will remove the password from memory, and any updates to the extension will clear the background context, so it won't ever be immediately available without the user entering it.
2. The password is sent around using [`runtime.sendMessage()`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/sendMessage), which only communicates to contexts created by the extension, and doesn't communicate to the content script.
3. All payments and invoice generation still require a prompt.
4. You can uncheck the "Don't ask me again" checkbox if you want and no messages will be sent anywhere.

### Steps to Test

1. Perform an action that requires a password (e.g. click Deposit). Leave the checkbox checked.
2. Close the popup, and perform that action again. Confirm you weren't asked for the password.
3. Open the menu dropdown and click "Lock account". Perform the action again and confirm you were asked for the password.
4. Enter the password, but uncheck the checkbox. Close the popup, perform the action again, and confirm you were asked for the password again.

### Screenshots

<img width="395" alt="screen shot 2018-12-28 at 1 59 16 pm" src="https://user-images.githubusercontent.com/649992/50525562-2c921500-0aaa-11e9-9e24-0412de2b7013.png">


<img width="277" alt="screen shot 2018-12-28 at 1 59 06 pm" src="https://user-images.githubusercontent.com/649992/50525557-24d27080-0aaa-11e9-9a39-ea697fe1c4ed.png">
